contracts/external/cw-fund-distributor/src/contract.rs

```diff
@@ -582,41 +565,44 @@ fn execute_redistribute_unclaimed_funds(
     DISTRIBUTION_HEIGHT.save(deps.storage, &distribution_height)?;
 
     // get performed claims of cw20 and native tokens
-    let performed_cw20_claims: HashMap<(Addr, Addr), Uint128> = CW20_CLAIMS
+    let performed_cw20_claims: Vec<Cw20ClaimEntry> = CW20_CLAIMS
         .range(deps.storage, None, None, Order::Descending)
-        .map(|native| native.unwrap())
         .collect();
-
-    let performed_native_claims: HashMap<(Addr, String), Uint128> = NATIVE_CLAIMS
+    let performed_native_claims: Vec<NativeClaimEntry> = NATIVE_CLAIMS
         .range(deps.storage, None, None, Order::Descending)
-        .map(|native| native.unwrap())
         .collect();
 
-    // subtract the performed claim amounts from
-    // balances available for claiming
-    performed_native_claims
-        .into_iter()
-        .for_each(|((_, denom), amount)| {
-            NATIVE_BALANCES
-                .update(deps.storage, denom, |bal| {
-                    bal.unwrap_or_default()
-                        .checked_sub(amount)
-                        .map_err(StdError::overflow)
-                })
-                .unwrap();
-        });
+    // subtract every performed claim from the available distributor balance
+    for entry in performed_cw20_claims {
+        let ((_, cw20_addr), amount) = entry?;
+        CW20_BALANCES.update(deps.storage, cw20_addr.clone(), |bal| {
+            // should never hit the None arm in theory
+            match bal {
+                Some(cw20_balance) => cw20_balance
+                    .checked_sub(amount)
+                    .map_err(ContractError::OverflowErr),
+                None => Err(ContractError::Std(StdError::NotFound {
+                    kind: cw20_addr.to_string(),
+                })),
+            }
+        })?;
+    }
 
-    performed_cw20_claims
-        .into_iter()
-        .for_each(|((_, cw20_addr), amount)| {
-            CW20_BALANCES
-                .update(deps.storage, cw20_addr, |bal| {
-                    bal.unwrap_or_default()
-                        .checked_sub(amount)
-                        .map_err(StdError::overflow)
-                })
-                .unwrap();
-        });
+    // subtract every performed claim from the available distributor balance
+    for entry in performed_native_claims {
+        let ((_, denom), amount) = entry?;
+        NATIVE_BALANCES.update(deps.storage, denom.clone(), |bal| {
+            // should never hit the None arm in theory
+            match bal {
+                Some(native_balance) => native_balance
+                    .checked_sub(amount)
+                    .map_err(ContractError::OverflowErr),
+                None => Err(ContractError::Std(StdError::NotFound {
+                    kind: denom.to_string(),
+                })),
+            }
+        })?;
```

Comment:
for some reason i never appreciated how sick the redistribution method on this is. as you continuously redistribute, the portion of funds retained for address' that are not claiming goes to zero 🤯

Reply:
Indeed, although I am not too happy with how slowly this reaches the state where all funds have been distributed.

Given a handful of people who miss the distribution and are unaware of the claim, these redistributions could become a little slow. And I would hate to see a contract with some non-zero amount of, say, juno, that would just not be worth the time to redistribute.

Couple of potential solutions/ideas:
1. introduce a time period/number of re-distributions after which all the _non-claimers_ would be excluded from further re-distributions
2. introduce a time period/number of re-distributions after which all the remaining funds get clawed-back into the treasury

Reply:
1. sounds like it would be fun to write